src/core/main/worker/worker_main.ts

```diff
@@ -516,30 +518,69 @@ function loadOrReloadPreparedContent(
   const {
     contentId,
     cmcdDataBuilder,
+    enableRepresentationAvoidance,
     manifest,
     mediaSource,
     representationEstimator,
     segmentSinksStore,
     segmentQueueCreator,
   } = preparedContent;
   const { drmSystemId, enableFastSwitching, initialTime, onCodecSwitch } = val;
-  playbackObservationRef.onUpdate((observation) => {
-    if (preparedContent.decipherabilityFreezeDetector.needToReload(observation)) {
-      handleMediaSourceReload({
-        timeOffset: 0,
-        minimumPosition: 0,
-        maximumPosition: Infinity,
+  playbackObservationRef.onUpdate(
+    (observation) => {
+      // Synchronize SegmentSinks with what has been buffered.
+      ["video" as const, "audio" as const, "text" as const].forEach((tType) => {
+        const segmentSinkStatus = segmentSinksStore.getStatus(tType);
+        if (segmentSinkStatus.type === "initialized") {
+          segmentSinkStatus.value.synchronizeInventory(observation.buffered[tType] ?? []);
+        }
       });
-    }
 
-    // Synchronize SegmentSinks with what has been buffered.
-    ["video" as const, "audio" as const, "text" as const].forEach((tType) => {
-      const segmentSinkStatus = segmentSinksStore.getStatus(tType);
-      if (segmentSinkStatus.type === "initialized") {
-        segmentSinkStatus.value.synchronizeInventory(observation.buffered[tType] ?? []);
+      const freezeResolution =
+        preparedContent.freezeResolver.onNewObservation(observation);
+      if (freezeResolution !== null) {
+        switch (freezeResolution.type) {
+          case "reload": {
+            log.info("WP: Planning reload due to freeze");
```

Comment:
Done in a way, though I was kind of conflicted:

  - There is a not-so visible dependency the `FreezeResolver` has on the segment sinks being synchronized first.
    I just tried to reimplement it so that `onNewObservation` takes the `segmentSinksStore` as a second argument so it is clearer - but by doing so it removed it from the constructor and IMO it may even be more important to indicate that a `FreezeResolver` needs to be re-created at each new `SegmentSinksStore` creation (because it makes no sense to compare between different SourceBuffers - which generally also means a new MediaSource - I think). So I did not do this modification at the end, even if it means that the dependency isn't clear.

    Though this dependency is why I prefer to keep them close in the same event listener.

  - After some attempt, I ended up writing those functions at the top level instead of an inner scope, even if it means that the freezing resolution one takes a lot of parameters. I generally prefer to put functions in top level because it indicate clearly the dependencies - even if sometimes the function signature does look better as a closure.

Reply:
And I re-checked but it seems that this dependency is not well taken into account in monothread. Looking at that right now

Reply:
Done for the second part

Reply:
Thanks, looks better